examples/fabric/tensor_parallel/model.py

```diff
@@ -0,0 +1,532 @@
+
+
+# Taken and modified from torchtitan
+# https://github.com/pytorch/torchtitan/blob/main/torchtitan/parallelisms/parallelize_llama.py
+def parallelize(model: Transformer, device_mesh: DeviceMesh) -> Transformer:
+    """Apply parallelisms and activation checkpointing to the model.
+
+    NOTE: The passed-in model preferably should be on meta device. Otherwise,
+    the model must fit on GPU or CPU memory.
+
+    """
+
+    dp_mesh = device_mesh["data_parallel"]
+    tp_mesh = device_mesh["tensor_parallel"]
+
+    if tp_mesh.size() > 1:
+        # 1. Parallelize the first embedding and the last linear proj layer
+        # 2. Parallelize the root norm layer over the sequence dim
+        # 3. Shard the first transformer block's inputs
+        # Parallelize the first embedding and the last linear out projection
+        plan = {
+            "tok_embeddings": RowwiseParallel(
+                input_layouts=Replicate(),
+            ),
+            "output": ColwiseParallel(input_layouts=Shard(1), output_layouts=Replicate()),
+            "norm": SequenceParallel(),
+            "layers.0": PrepareModuleInput(
+                input_layouts=(Replicate(), None),
+                desired_input_layouts=(Shard(1), None),
+                use_local_output=True,
+            ),
+        }
+        model = parallelize_module(model, tp_mesh, plan)
+
+        # Parallelize each transformer block
+        for transformer_block in model.layers:
+            plan = {
+                "attention": PrepareModuleInput(
+                    input_layouts=(Shard(1), None),
+                    desired_input_layouts=(Replicate(), None),
+                ),
+                "attention.wq": ColwiseParallel(),
+                "attention.wk": ColwiseParallel(),
+                "attention.wv": ColwiseParallel(),
+                "attention.wo": RowwiseParallel(output_layouts=Shard(1)),
+                "attention_norm": SequenceParallel(),
+                "feed_forward": PrepareModuleInput(
+                    input_layouts=(Shard(1),),
+                    desired_input_layouts=(Replicate(),),
+                ),
+                "feed_forward.w1": ColwiseParallel(),
+                "feed_forward.w2": RowwiseParallel(output_layouts=Shard(1)),
+                "feed_forward.w3": ColwiseParallel(),
+                "ffn_norm": SequenceParallel(),
+            }
+
+            # Adjust attention module to use the local number of heads
+            attn_layer = transformer_block.attention
+            attn_layer.n_heads = attn_layer.n_heads // tp_mesh.size()
+            attn_layer.n_kv_heads = attn_layer.n_kv_heads // tp_mesh.size()
+
+            # Apply the plan for the current transformer block
+            parallelize_module(transformer_block, tp_mesh, plan)
+
+    if dp_mesh.size() > 1:
+        assert dp_mesh.ndim == 1  # Hybrid-sharding not supported
+
+        mp_policy = MixedPrecisionPolicy(param_dtype=torch.bfloat16, reduce_dtype=torch.float32)
```

Comment:
Isn't this meant to be encapsulated by fabric?

Reply:
Since the user modifies the model directly here, probably not. In any case not for this iteration.

Reply:
Worth leaving a comment then because as a user I would be confused about the impact of doing this and setting `precision="something"` in Fabric

Reply:
Right, I added a comment. We will need to do something about precision soon, let's brainstorm this